Topics/Tech_Stacks/oopcpp.md

```diff
@@ -0,0 +1,200 @@
+C++ can be a confusing language. Pointers, ownership, memory management -- there is a lot going on in even a simple C++ program and it is often difficult for beginners to approach. As students entering the workforce we want to learn how to write industry grade C++ that would impress our interviewers, coworkers, and employers. This guide gives an introduction to doing just this, covering some of the most important topics in object-oriented design in C++.
+
+### So what is OOP?
+Object-oriented programming or OOP are the fundamental principles that most modern languages are optimized to follow. We can summarize OOP with 4 primary concepts.
+* **Encapsulation** is the principle of bundling data and the methods that operate on that data under a common slass. This allows us to hide the internal details of an object and restrict access to its state.
+* **Inheritance** is the mechanism that allows a class to inherit the properties and behaviors of another, known as the parent class. This promotes code reusability and facilitates the creation of a hierarchical structure for classes.
+* **Polymorphism** means "many forms" and is concept that allows objects of different classes to be treated as objects of a common parent class, enabling code to work with objects of various types in a uniform way. In particular, mechanisms like method overloading, defining multiple methods with the same name but different parameters, and method overriding, providing a specific implementation of a parent classes' method in a child class, make polymorphism possible.
+* **Abstraction** involves simplifying complex systems by modeling classes based on the essential properties and behaviors that they share, focusing on what an object does rather than how it does it. This make it easy for code to be utilized by different developers and reduces complexity by hiding unnecessary details.
+
```

Comment:
It would be really helpful if you could add more details to differentiate the concepts. Some OOP concepts are very similar to each other and can be quite confusing to beginners.

Reply:
Hey David, I think this is pretty difficult to achieve. I'm not sure there is much more that I can add to this in terms of detail since this is a very simple and fundamental concept.